controller/src/main/java/org/jboss/as/controller/operations/global/QueryOperationHandler.java

```diff
@@ -0,0 +1,261 @@
+                                    && context.getResult().isDefined() // exclude empty model nodes
+                                    ){
+                                try {
+                                    ModelNode reduced = reduce(context.getResult(), operation.get(SELECT));
+                                    context.getResult().set(reduced);
+                                } catch (OperationFailedException e) {
+                                    context.getFailureDescription().set(e.getMessage());
+                                    context.setRollbackOnly();
+                                }
+                            }
+
+                        }
+                    }
+            );
+
+        }
+
+        private static boolean matchesFilter(final ModelNode resource, final ModelNode filter, final Operator operator) throws OperationFailedException {
+            boolean isMatching = false;
+            List<Property> filterProperties = filter.asPropertyList();
+            List<Boolean> matches = new ArrayList<>(filterProperties.size());
+
+            for (Property property : filterProperties) {
+
+                final String filterName = property.getName();
+                final ModelNode filterValue = property.getValue();
+
+                boolean isEqual = false;
+
+                if(!filterValue.isDefined() || filterValue.asString().equals(UNDEFINED))  {
+                    // query for undefined attributes
+                    isEqual = !resource.get(filterName).isDefined();
+                }  else {
+
+                    final ModelType targetValueType = resource.get(filterName).getType();
+
+                    try {
+                        // query for attribute values (throws exception when types don't match)
+                        switch (targetValueType) {
+                            case BOOLEAN:
+                                isEqual = filterValue.asBoolean() == resource.get(filterName).asBoolean();
+                                break;
+                            case LONG:
+                                isEqual = filterValue.asLong() == resource.get(filterName).asLong();
+                                break;
+                            case INT:
+                                isEqual = filterValue.asInt() == resource.get(filterName).asInt();
+                                break;
+                            case DOUBLE:
+                                isEqual = filterValue.asDouble() == resource.get(filterName).asDouble();
+                                break;
+                            default:
+                                isEqual = filterValue.equals(resource.get(filterName));
+                        }
+                    } catch (IllegalArgumentException e) {
+                        throw ControllerLogger.MGMT_OP_LOGGER.validationFailedCouldNotConvertParamToType(filterName, targetValueType, ModelDescriptionConstants.QUERY);
```

Comment:
This method throws IllegalArgumentException, not OperationFailedException. But your catch block at L144 is counting on catching OperationFailedException.

I was testing the WFCORE-442 fix and suddenly /path=*:query(where={read-only=-1}) wouldn't fail properly any more. This is why.